shared/src/main/kotlin/com/egm/stellio/shared/model/ExpandedEntity.kt

```diff
@@ -1,68 +1,67 @@
 package com.egm.stellio.shared.model
 
+import arrow.core.Nel
+import arrow.core.Validated
+import arrow.core.extensions.nonemptylist.semigroup.semigroup
+import arrow.core.extensions.validated.applicative.applicative
+import arrow.core.fix
+import arrow.core.invalid
+import arrow.core.orNull
+import arrow.core.valid
+import arrow.core.validNel
 import com.egm.stellio.shared.util.AttributeType
 import com.egm.stellio.shared.util.NgsiLdParsingUtils
 import com.egm.stellio.shared.util.NgsiLdParsingUtils.NGSILD_DATASET_ID_PROPERTY
 import com.egm.stellio.shared.util.NgsiLdParsingUtils.NGSILD_ENTITY_CORE_MEMBERS
 import com.egm.stellio.shared.util.NgsiLdParsingUtils.NGSILD_ENTITY_ID
+import com.egm.stellio.shared.util.NgsiLdParsingUtils.NGSILD_ENTITY_TYPE
 import com.egm.stellio.shared.util.NgsiLdParsingUtils.NGSILD_GEOPROPERTY_TYPE
 import com.egm.stellio.shared.util.NgsiLdParsingUtils.NGSILD_PROPERTY_TYPE
 import com.egm.stellio.shared.util.NgsiLdParsingUtils.NGSILD_RELATIONSHIP_TYPE
+import com.egm.stellio.shared.util.NgsiLdParsingUtils.expandValueAsListOfMap
+import com.egm.stellio.shared.util.NgsiLdParsingUtils.expandValueAsMap
+import com.egm.stellio.shared.util.NgsiLdParsingUtils.isAttributeOfType
+import com.egm.stellio.shared.util.NgsiLdParsingUtils.isValidAttribute
 import com.egm.stellio.shared.util.extractShortTypeFromExpanded
 import com.github.jsonldjava.core.JsonLdOptions
 import com.github.jsonldjava.core.JsonLdProcessor
 
-class ExpandedEntity private constructor(
+// TODO we define a list of errors, this may not be necessary as we ultimately return the reason string to the consumers
+sealed class EntityParsingError(val reason: String) {
+    object MissingIdError : EntityParsingError("The provided NGSI-LD entity does not contain an id property")
+    object MissingTypeError : EntityParsingError("The provided NGSI-LD entity does not contain a type property")
+    class MixedAttributeInstancesError(reason: String) : EntityParsingError(reason)
+    class NonUniqueDatasetIdError(reason: String) : EntityParsingError(reason)
+    class MoreThanOneDefaultInstanceError(reason: String) : EntityParsingError(reason)
+
+    override fun toString(): String = reason
+}
+
+data class JsonLdExpandedEntity(
     val rawJsonLdProperties: Map<String, Any>,
     val contexts: List<String>
 ) {
-    companion object {
-        operator fun invoke(attributes: Map<String, Any>, contexts: List<String>): ExpandedEntity {
-            if (!attributes.containsKey("@id"))
-                throw BadRequestDataException("The provided NGSI-LD entity does not contain an id property")
-
-            if (!attributes.containsKey("@type"))
-                throw BadRequestDataException("The provided NGSI-LD entity does not contain a type property")
-
-            return ExpandedEntity(attributes, contexts)
-        }
-    }
-
+    // FIXME kinda hacky but we often just need the id ... how can it be improved?
```

Comment:
Move the part in which we need the id after the conversion to an expanded entity could be a solution, no?